Port competition to rails performance improvements

timhabermaas opened the pull request:
## Todo
- [ ] Fix the model spec I broke
- [x] Add view caching (how?)

jfly committed:
Ported competition infos to ERB
Also changed most links to point to /competitions/:competition_id url
instead of old PHP competition.php file. #279

jfly committed:
Only show links to /register and the list of competitors up until
results are uploaded. At any time before then, I imagine it would be
nice to check who was planning to compete in the competition, rather
than knowing nothing at all. We also now have alerts that inform you
about ongoing or over competitions that don't have results uploaded yet. #279

jfly committed:
Added nginx rules to redirect requests for old PHP competitions page to
new Rails page. I'm a little afraid to make these permanent redirects, so
it's just a 302 for now. #279. @FatBoyXPC, I hope this makes you happy.

jfly committed:
Added a migration to convert all processLinks style columns into
markdown style links.

pedrosino committed:
Continuing the new competitions index.

pedrosino committed:
Seeds more competitions (41) and trying an anchor to "center" the page on today.

pedrosino committed:
Added the selectors to the competitions page. Continent is dumb, though.

jfly committed:
Finished porting competition.php to Rails! This fixes #279.

jfly committed:
Removing remaining references to c.php/competition.php and deleted the
old PHP code. This really fixes #279.

pedrosino committed:
Selectors now work properly, including the continents.

pedrosino committed:
Moving the new_index to index and updating the link on the main menu.

pedrosino committed:
Added a checkmark indicating if results are posted.

jfly committed:
Don't look up the person for each Result, instead use the personName
field we already have.

pedrosino committed:
Forgot to update the competitions_controller to allow non-authenticated users to access /competitions.

pedrosino committed:
Fixed some broken stuff, and changed the way to "center" the competition index on today.

jfly committed:
Rename _competition_results to _results_table, and change it to use our
own new wca_table helper, rather than wca_table_for, which was being
*super* slow. Also expanded database seeds to generate lots of results
for competitions, rather than only a few. #279.

jfly committed:
Ensure that solves always returns an array of length 5, which makes
styling easier. Also removed some vestigial table-for columns.

jfly committed:
Default to nowrap for greedy-last-column.

jfly committed:
Make sure solves less than 1 second show 0 in the seconds area.

jfly committed:
Fixed broken test that was supposed to look at organizers, not delegates.

jfly committed:
Removed vestigial wca_table_for_options argument.

jfly committed:
Add /lib to the autoload directory and remove corresponding requires.
This fixes #324.

pedrosino committed:
Fixed a lot of stuff.

pedrosino committed:
Added some more styling and bootstrap overrides.

pedrosino committed:
Using jquery to move to #today anchor, and changed link colors as per Lucas' comment.

jfly committed:
Redirect old php links to registration list and registration form to new
Rails routes as per @Luis-J-Ianez's request.

pedrosino committed:
Fixed styling, added the map and some final touches.

pedrosino committed:
Fixed the map (no more locationpicker) and changed the marker colors to differentiate future and past competitions.

pedrosino committed:
Select only competition with showAtAll = true.

pedrosino committed:
Default option for 'years' (current).

pedrosino committed:
Fixed the @closest_index calculation.

pedrosino committed:
Added results_uploaded? to competition.

Conflicts:
	WcaOnRails/app/views/competitions/index.html.erb

pedrosino committed:
Using rails to format the date for markers, and added clustering.

pedrosino committed:
Fixed the "today" styling and positioning.

pedrosino committed:
Added MarkDown to the venue cell.

pedrosino committed:
Re-fixing the "today line" position.

pedrosino committed:
Some more styling.

pedrosino committed:
Merge branch 'master' of http://github.com/pedrosino/worldcubeassociation.org

Conflicts:
	WcaOnRails/app/assets/stylesheets/competitions.scss
	WcaOnRails/app/assets/stylesheets/navbar-static-top.scss
	WcaOnRails/app/views/competitions/my_competitions.html.erb

pedrosino committed:
Changed buttons style and added floatThead.

pedrosino committed:
Moved the selector inside the floatin thead.

pedrosino committed:
Seeding more competitions and fixing some styles.

pedrosino committed:
Trying without the results...

pedrosino committed:
Moved the form into a partial and made the map bigger.

pedrosino committed:
No wca_highlighting, seems it could be one of the causes for bad performance.

pedrosino committed:
Added some checks for invalid form params.

pedrosino committed:
Actually, we may not need param checks...also added a message when no competition is found.

pedrosino committed:
Merge branch 'master' of https://github.com/pedrosino/worldcubeassociation.org

Conflicts:
	WcaOnRails/app/controllers/competitions_controller.rb

pedrosino committed:
Many fixes from Jeremy and Tim comments.

pedrosino committed:
Using is_over? to separate future and past competitions.

pedrosino committed:
Added the optgroup for regions (finally).

pedrosino committed:
Added a Continent model to be used for the competitions index.

pedrosino committed:
Nitpick map{ -> map {

pedrosino committed:
More fixes from Jeremy's comments.

pedrosino committed:
Merge branch 'master' of http://github.com/pedrosino/worldcubeassociation.org

pedrosino committed:
Fixed the selects and the map and scroll functions. Also removed bootstrap override -> will move to a separate PR.

pedrosino committed:
Overriding bootstrap's table striped colors.

pedrosino committed:
Fixed the years array sorting.

pedrosino committed:
Removed the results_uploaded? check, because it makes the view 2x slower...

pedrosino committed:
Moved map and scroll functions to application.js

pedrosino committed:
More fixes, added labels for the selectors and put the buttons inside a single div.

pedrosino committed:
Merge branch 'master' of http://github.com/pedrosino/worldcubeassociation.org

Conflicts:
	WcaOnRails/app/assets/javascripts/application.js

pedrosino committed:
Search is now case-insensitive.

pedrosino committed:
Fixed the buttons position and some indentation.

pedrosino committed:
Buttons side by side and redirects.

pedrosino committed:
No floating on mobile, and trying to fix the competitions map.

pedrosino committed:
Trying something with the map.

pedrosino committed:
Little hack to set the map size. Not perfect on mobile, but still looks good.

pedrosino committed:
Moved the map size to a function, and also call it when the window is resized.

pedrosino committed:
Creating teams table, model and controller.

pedrosino committed:
Slightly different schema.

pedrosino committed:
Created team_members table and nested them inside teams/edit.

pedrosino committed:
Proper labels for members, leader selectable from members.

jfly committed:
Attempting to switch from mandrill to Amazon SES. Fingers crossed that
this actually works.

jfly committed:
Attempting to switch from mandrill to Amazon SES. Fingers crossed that
this actually works.

jfly committed:
Revert "Attempting to switch from mandrill to Amazon SES. Fingers crossed that"

This reverts commit c8c2f10c562f340abef733f77fc4b15b8717c108.

coder13 committed:
Added delete button for users' registeration. Fixes #269.

coder13 committed:
fixed tests now that competitors can delete their registration. #269

jfly committed:
Workaround for #409.

jfly committed:
Better styling of simple_form checkboxes to simplify our competitions
edit page. Also changed guests_enabled checkbox to a painfully explicit
set of 2 radio buttons inspired by CubingUSA:
https://github.com/cubing/worldcubeassociation.org/issues/286#issuecomment-168251174
(fixes #404).

pedrosino committed:
Don't tab on datetime_picker hide.

jfly committed:
Merge pull request #410 from jfly/issue-404

Better styling of simple_form checkboxes to simplify our competitions edit page

pedrosino committed:
Using keepOpen instead of auto tabbing.

pedrosino committed:
Team_leader is now a bool in the team_member table.

pedrosino committed:
Removed leader from team form and added to team_member form.

coder13 committed:
added tests for users's deleting their registration

jfly committed:
Use notifications@, rather than no-reply@ everywhere. I opted to stick
with notifications@ since most of our registration related emails have
been notifications@, and I don't want to break email filters that people
may have set up.

coder13 committed:
added missing periods to emails

pedrosino committed:
Trying simple_fields_for.

jfly committed:
Merge pull request #408 from coder13/learning

Added delete button to users' registration

pedrosino committed:
Fixed the controller params and added as: :date_picker.

jfly committed:
Remove contact form for WRC in favor of a direct email address.

pedrosino committed:
Fixed some things from comments.

jfly committed:
Be a data liberal: it's more likely we want to render nothing when
something in the database is nil, rather than crash.

pedrosino committed:
Migration to fill the tables.

pedrosino committed:
Fixed the migration.

pedrosino committed:
Running the migration.

jfly committed:
Merge pull request #411 from pedrosino/dates

Don't tab on datetime_picker hide.

pedrosino committed:
Added autocomplete and trying a better styling.

jfly committed:
Ported @pochmann's 6028b750c89cf79649a5dbcf2696467bf2cf6e7e to our new
Rails system.

jfly committed:
Cleaned stuff up so it's clear that everything is required.

jfly committed:
Allow newcomers to fill in birthdate, gender, citizenship when signing
up for an account. This is a step towards #412.

jfly committed:
Oops! da72b51ae41376c584da7eb524a857892375d772 broke google analytics.
MY BAD.

pedrosino committed:
Better styling for the member form.

pedrosino committed:
Removed the required asterisk and changed the user id input style.

jfly committed:
Workaround for footer where it was always inducing a vertical scrollbar.

jfly committed:
Optimize account creation, especially for people who have actually
competed before.
We now ask if you've competed before, and if you have, we only
ask for your WCA ID, since we can figure out the rest of your
information from that. Note that this doesn't actually give you that WCA
ID, instead it sets up a WCA ID claim (but we don't bother notifying or
emailing your delegate until you've confirmed your email address).

If you have not competed before, there isn't much we can do... we just
ask you for your dob, gender, and citizenship.

This fixes #412.

jfly committed:
Add a much needed border to the footer.

pedrosino committed:
Removing old teams columns.

pedrosino committed:
Fixed permissions and almost all tests.

jfly committed:
Fixed bugs with account creation and WCA ID claiming when there are
validation errors. That was surprisingly tricky.
Added a little more space between the body and the footer.

jfly committed:
Don't show WCA ID not existing message if WCA ID is blank.

jfly committed:
Cleanup as per discussion on
https://groups.google.com/d/msg/plataformatec-simpleform/aHAtI3pONJo/B_YDb9u1AgAJ

pedrosino committed:
Removed team leader validation and some of the tests.

jfly committed:
Merge pull request #413 from jfly/issue-412

Optimize account creation flow

pedrosino committed:
Trying Jim's design

pedrosino committed:
Finished the new design.

pedrosino committed:
Remove scrollToCenter and adjusted the z-index thing.

pedrosino committed:
Using row and col-md-* instead.

pedrosino committed:
Other way to put the dates.

pedrosino committed:
Different style for in-progress competitions.

pedrosino committed:
Anchor links?

jfly committed:
Merge pull request #371 from pedrosino/bootstrap

Overriding bootstrap's table striped colors.

pedrosino committed:
Fixed the something_team? methods.

pedrosino committed:
Fixed team_member? method and added was_team_member?

pedrosino committed:
Removed vestigial stuff, added comments and fixed some other things.

pedrosino committed:
Trying bootstrap-tookit for the scrolling div.

pedrosino committed:
Yet another design. Made all lists col-md-12.

pedrosino committed:
Trying to correctly position the anchors.

pedrosino committed:
Using <a> instead of <span> seems to work, except if you click the anchor before scrolling.

pedrosino committed:
Using javascript to jump to the anchors.

flee135 committed:
Don't allow people to delete accepted registrations. #417

flee135 committed:
Fixed typo in popup box

flee135 committed:
Fixes #417

pedrosino committed:
Added visual indication of past members.

pedrosino committed:
Cannot demote oneself and a few controller tests.

jfly committed:
Merge pull request #420 from flee135/master

Changes for fixing #417

jfly committed:
Add Uzbekistan as a country.

pedrosino committed:
More tests for teams controllers.

pedrosino committed:
One more test - cannot demote oneself.

jfly committed:
Added missing files for forum to work. Thanks @gagou9!

jfly committed:
Use rack-cors library to respond to CORS OPTIONS requests as well as
just simple GET requests.

pedrosino committed:
Fixed :bla => ble to bla: ble.

pedrosino committed:
Changed the team_member? methods and removed old stuff from user model.

pedrosino committed:
Some changes as suggested by Jim.

jfly committed:
floatThead broke our select all/none checkbox. Workaround floatThead by
ignoring the dummy table it creates, and using the getRowGroups() method
instead of looking up the thead ourselves. This fixes #423.

jfly committed:
Ported registration selectable table from table-for to our own, simple
wca_table method. This was our last use of table-for, so I was able to
remove the gem entirely. This fixes #331.

pedrosino committed:
Fixed the tests, the team_member? method and removed old code.

pedrosino committed:
Added 'edit teams' to the admin panel, and updated statistics icon.

pedrosino committed:
Removed the (empty) destroy method from teams_controller.

pedrosino committed:
Some changes to factories, methods and application.js.

pedrosino committed:
Using variables for styles and avoiding multiple listeners for blur.

pedrosino committed:
One more test and validation.

pedrosino committed:
Some css changes and keeping the navbar on mobile too.

pedrosino committed:
Merge branch 'master' of http://github.com/pedrosino/worldcubeassociation.org

Conflicts:
	WcaOnRails/app/helpers/application_helper.rb

pedrosino committed:
Moved team creation to happen at each test.

pedrosino committed:
Changing the format_team_member method.

pedrosino committed:
Using bootstrap variables and added a border to the #simple-nav.

pedrosino committed:
Using $(#simple-nave).height().

pedrosino committed:
Placeholder for the nav bar, to make scrolling work correctly.

pedrosino committed:
Fixed css.

pedrosino committed:
Static_pages_helper spec created.

pedrosino committed:
Ordering by team_leader and user.name.

pedrosino committed:
Improving the helper test.

jfly committed:
Don't spoof emails. This fixes #433.

pedrosino committed:
Links to markdown and better alignment.

pedrosino committed:
Merge branch 'master' of https://github.com/pedrosino/worldcubeassociation.org

Conflicts:
	WcaOnRails/app/assets/javascripts/application.js

jfly committed:
Merge pull request #319 from pedrosino/master

New competitions index.

jfly committed:
Only style p tags inside of .competition-list, not all p tags
everywhere.

jfly committed:
Gah! We were already inside of a .competition-list, whoops.

pedrosino committed:
Oops, some css was broken and also fixed competition.country_name.

pedrosino committed:
Merge pull request #435 from pedrosino/master

Fixing my_competitions page, as per reported on e-mail group.

pedrosino committed:
Fixing alignments as per Antonio's comment.

jfly committed:
Merge pull request #436 from pedrosino/master

Fixing alignments as per Antonio's comment.

jfly committed:
Fixed bug where users who have never competed before would get an error
message when confirming their account.

jonatanklosko committed:
Fix #432.

jfly committed:
Show a nice alert for accounts that are waiting for a delegate to
confirm their unconfirmed WCA ID.

jfly committed:
Merge pull request #437 from jonatanklosko/issue-432

Positions within psych sheet respects an occurrence of a tie.

jonatanklosko committed:
Fix issues connected with database setup.

jfly committed:
Merge pull request #439 from jonatanklosko/fix-seeds

Fix issues connected with database setup

jonatanklosko committed:
Add a hourglass icon based on a competition's start time.

jonatanklosko committed:
Add a blank line separating the years on the list.

jonatanklosko committed:
Change the captions for the past competitions table.

jonatanklosko committed:
Fix displaying competitions in future years when current year is selected.

jfly committed:
Update to latest version of floatThead with fix for
https://github.com/mkoryak/floatThead/issues/263.

Also fixed bug where we were generating <td>s inside of a <thead>
(fixes #429)

jfly committed:
Merge pull request #442 from jfly/latest-floatthead

Update to latest version of floatThead

jonatanklosko committed:
Add national flags.

pedrosino committed:
Add city and country to results announcement.

pedrosino committed:
Changed the post_results test.

jfly committed:
Merge pull request #440 from jonatanklosko/comp-index-improvements

Competitions index improvements

jfly committed:
Merge pull request #446 from pedrosino/master

Add city and country to results announcement.

jonatanklosko committed:
Fix multiday competitions to fit nicely.

timhabermaas committed:
Log user id in case of exception

Fixes #383.

jfly committed:
Merge pull request #453 from timhabermaas/log-user-id-in-case-of-exception

Log user id in case of exception

jfly committed:
Merge pull request #452 from jonatanklosko/hotfix-issue-451

Fix multiday competitions to fit nicely.

viroulep committed:
Bump TNoodle official version

Next TNoodle official version is 0.11.1.

jonatanklosko committed:
Make registrations table sortable.

jonatanklosko committed:
Change the default order for the first click on sort to be ascending.

jonatanklosko committed:
Refactor the code and make visible that the table is sorted by name by default.

jonatanklosko committed:
Make sorting always ascending on the first click.

pedrosino committed:
Selects all upcoming competitions when year==current.

jfly committed:
Merge pull request #459 from pedrosino/master

Selects all upcoming competitions when year==current.

jonatanklosko committed:
Make registrations table sortable on the edit page. Also benefit more from bootstrap-table by using its checkboxes functionality.

jonatanklosko committed:
Refactor the code.

jonatanklosko committed:
Add ability to search for competitions matching specific events.

jonatanklosko committed:
Add handy helper for pages specific JavaScript code.

jonatanklosko committed:
Add filtering out competitions that doesn't have at least one of the selected events.

pedrosino committed:
More countries, fake city and venue names for seeds.

viroulep committed:
Update the readme so that it links to the Wiki.

jfly committed:
Merge pull request #455 from jonatanklosko/sortable-registrations

Make registrations table sortable.

jfly committed:
Merge pull request #454 from viroulep/tnoodle_0.11.1

Bump TNoodle official version

jonatanklosko committed:
Move users management page from wice_grid to bootstrap-table in order to sort/search/paginate asynchronously.

pedrosino committed:
Random countries from Country model.

pedrosino committed:
Using sample.id instead.

jfly committed:
Merge pull request #463 from pedrosino/seeds

More countries, fake city and venue names for seeds.

jfly committed:
Merge pull request #465 from jonatanklosko/pages-specific-javascript-helper

Add handy helper for pages specific JavaScript

timhabermaas committed:
Memoize result methods

`sorted_solves_with_index` and `solves is used by various `Result`
methods. Caching the computed values cuts down on the required
allocations.

Response time (US Nationals 2015): 1.75s -> 1.4s
Allocations (String): 114264 -> 96904
Allocations (Array): 85568 -> 46271
Allocations (SolveTime): 11721 -> 3038

timhabermaas committed:
Freeze constant Strings in SolveTime

Response time (US Nationals 2015): 1.4s -> 1.4s
Allocations (String): 96904 -> 83082

timhabermaas committed:
Replace content_tag helper with HTML tag

Also: Use String#<< instead of Array#<< + Array#join for creating class
string.

Response time (US Nationals 2015): 1.4s -> 1.0s
Allocations (String): 96904 -> 47593
Allocations (Array): 46271 -> 28473

jonatanklosko committed:
Change bootstrap-table pagination description

pedrosino committed:
Using cellName for the competitions index.

jonatanklosko committed:
Refactor the code.

jfly committed:
Merge pull request #467 from pedrosino/master

Using cellName for the competitions index.

jonatanklosko committed:
Update footer in edit registrations table.

jfly committed:
Merge pull request #469 from jonatanklosko/hotfix-issue-468

Update footer in edit registrations table.

jfly committed:
Merge pull request #461 from jonatanklosko/search-for-specific-events

Add ability to search for competitions matching specific events.

jonatanklosko committed:
Entirely remove wice_grid.

jonatanklosko committed:
Refactor the code.

jfly committed:
Merge pull request #466 from jonatanklosko/user-management-table

Move users management page from wice_grid to bootstrap-table

jonatanklosko committed:
Add a tooltip with additional time information on hourglass hover.

FatBoyXPC committed:
Merge pull request #476 from jonatanklosko/hourglasses-tooltips

Add a tooltip with additional time information on hourglass hover.

jonatanklosko committed:
Fix registrations with updated_at set to nil to not cause an error.

timhabermaas committed:
Don't go through ActiveRecord when getting results

Profiling the rendering of _results_table.html.erb showed quite some
time was spent in `ActiveRecord#read_attribute`. So, I load the results
using raw SQL and instantiate a PORO. The code definitely got uglier,
but the performance gains are worth it IMO. Not using ActiveRecord led
to a 40% performance improvement:

Response time (US Nationals 2015): 1.0s -> 600ms
Allocations (String): 47593 -> 41515
Allocations (Array): 28473 -> 26303

timhabermaas committed:
Convert HTML comments to Ruby comments

This saves a few bytes...

jfly committed:
Merge pull request #477 from jonatanklosko/fix-registrations-without-updated-at

Fix registrations with updated_at set to nil to not cause an error

jonatanklosko committed:
Fix issue with event ids passed as a hash.

jfly committed:
Merge pull request #478 from jonatanklosko/fix-uri-with-events

Fix issue with event ids passed as a hash

timhabermaas committed:
Get rid of N+1 queries for person on by_person competition page

The average response time for by_person (US Nationals 2015) went
from 2.3s to 1.0s.

timhabermaas committed:
Skip ActiveRecord for by_person results as well

Response Time: 1.0s -> 0.7s

timhabermaas committed:
Use LightResult consistently for all competition results

Also: Get rid of building SQL using String interpolation by using
Relation#to_sql.

timhabermaas commented:
@jfly I'm "done" with the obvious optimizations. The majority of the time is now spent in actually rendering the views (e.g. `ActionView::OutputBuffer#<<`). ~6% is still spent in `SolveTime#clock_format`, but this accounts for max 0.06s per request.

Here are the current response times (90% percentile measured with `ab -n 30` against the production environment) for US Nationals 2015:

| Page | Before (`port-competition-to-rails`) | After (`port-competition-to-rails-performance-improvements`) |
| --- | --: | --: |
| `/` | 45ms | 40ms |
| `/results/all` | 1704ms | 703ms |
| `/results/by_person` | 3110ms | 780ms |
| `/results/podiums` | 64ms | 52ms |

The next step would be to introduce (fragment) caching. When caching the main content of the competition view the response times drop to <= 150ms. 150ms for large competitions isn't really amazing, but there's probably not much we can do about it: The HTML is just huge (~3MB) and Rails still needs to embed it in the current page.

In order to do that I need a proper cache key, though. Something like `"USNationals2015-results_all-#{competition.results.order(:updated_at).last.updated_at}"` would be perfect. We don't have an `updated_at` column, though. Instead of introducing one and updating some PHP code I want to add an [auto-updating column](http://dev.mysql.com/doc/refman/5.7/en/timestamp-initialization.html) which we can later replace with ActiveRecord's `updated_at`. Any objections?

timhabermaas committed:
Clean up implementation

jfly committed:
Fixed bug in migration. We can't just set created_at to
registration.competition.registration_open, because not all competitions
actually *have* a registration_open.

jfly committed:
Possible workaround for broken caching. See
https://github.com/travis-ci/travis-ci/issues/4608

jfly commented on WcaOnRails/app/models/light_result.rb:
```diff
@@ -0,0 +1,108 @@
+
+  private def sorted_solves_with_index
+  @sorted_solves_with_index ||= solves.each_with_index.reject { |s, i| s.skipped? }.sort
+end
```
nit: indentation of this file looks screwy at and past this point

timhabermaas replied:
Thanks, NeoVim can't seem to handle the inline `private`...

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -473,6 +473,12 @@ def events_with_rounds_with_results
 
+  private def light_results_from_relation(relation)
+    ActiveRecord::Base.connection.execute(relation.to_sql).each(as: :hash).map do |r|
+      LightResult.new(r)
```
Interesting. Can you explain the reason for creating a LightResult rather than a regular Result?

timhabermaas replied:
I've "explained" it here: https://github.com/jfly/worldcubeassociation.org/pull/1/commits/743e0b1218d0bfed0bd2e988430348f8c6e7142a I didn't dive into ActiveRecord internals in order to figure out why exactly it's so slow, but creating and using thousands of ActiveRecord models has never been particularly fast.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -441,15 +441,15 @@ def results_uploaded?
+    light_results_from_relation(results.where(roundId: Round.final_rounds.map(&:id)).where("pos = 1")).sort_by { |r| r.event.rank }
   end
 
   def persons_with_results
```
nit: I believe that this function should be renamed. How about the more explicit `person_names_with_results`?

jfly commented:
Thanks so much for taking this on, @timhabermaas! There are some neat ruby internals here.

The idea of an autoupdating column that doesn't require any PHP work sounds really slick, and I'd be all for it if you can do it!

jonatanklosko committed:
Change searching for selected events to be strict.

jfly committed:
Merge pull request #480 from jonatanklosko/strict-events-matching

Change searching for selected events to be strict.

jonatanklosko committed:
Add simple flag imitations for competitions taking place in multiple countries.

pedrosino committed:
Merge pull request #482 from jonatanklosko/fix-multiple-countries-flags

Add simple flag imitations for competitions taking place in multiple countries

timhabermaas commented:
Fun fact: Caching the podiums and winners table is actually slower than always rendering them.

timhabermaas committed:
Cache competition results

I didn't cache the podiums and winners pages because caching them was
actually slower than not caching them at all (~65ms vs ~45ms). The
results/by_person and results/all pages are cached using a cache key
like:

    USNationals2015/2016-04-01T20:32:14Z/by_person/d9fb66b120b61f4...

The last part is generated by Rails and is the MD5 hash of the template
itself. Therefore it isn't technically necessary to include the string
"by_person", but I liked the explicity of it and I feel it's less
confusing for people not terrible familiar with fragment caching.

The second part is the `updated_at` column of the newest result for that
competition. There was no such column and adding one the Rails way
would've required us to update the PHP code, so I used a self-updating
column as described in:
http://gusiev.com/2009/04/update-and-create-timestamps-with-mysql/
As explained in the blog post you can't have two columns with `DEFAULT
NOW()` in MySQL without changing the `INSERT` code. That's why I didn't
add the `created_at` column as well.

The average response times for USNationals2015 dropped from ~700ms to
~150ms.

timhabermaas commented:
> The idea of an autoupdating column that doesn't require any PHP work sounds really slick, and I'd be all for it if you can do it!

It was actually pretty straightforward: https://github.com/jfly/worldcubeassociation.org/pull/1/commits/50451953c0bc12bec0a6d5bb912b9820a41e67b5

See the commit message for some details.

timhabermaas commented:
Btw. I'll close this PR (after I've fixed the specs and integrated your comments) and open one against `cubing/worldcubeassociation/master`. Unless you want to merge this as it is and continue working on your branch?

timhabermaas committed:
Fix code style

jfly commented:
If you're up for pushing this over the finish line, I'd really appreciate it. I'm swamped just trying to keep up with email lately!

timhabermaas commented:
Alright. Work will be continued in https://github.com/cubing/worldcubeassociation.org/pull/485
